src/ray/raylet/node_manager.cc

```diff
@@ -421,9 +433,46 @@ void NodeManager::DestroyWorker(std::shared_ptr<WorkerInterface> worker,
   KillWorker(worker);
 }
 
+void NodeManager::HandleJobSubmitted(const JobID &job_id, const JobTableData &job_data) {
+  RAY_LOG(DEBUG) << "HandleJobSubmitted " << job_id;
+  RAY_CHECK(!job_data.is_dead());
+  // Notify job manager that the job is submitted.
+  if (!job_manager_.OnJobSubmitted(std::make_shared<JobTableData>(job_data))) {
+    auto local_job_data = job_manager_.GetJobData(job_id);
+    RAY_CHECK(local_job_data != nullptr);
+    RAY_LOG(WARNING) << "Failed to handle job submitted event, local state is "
+                     << rpc::JobTableData_JobState_Name(local_job_data->state())
+                     << ", received state is "
+                     << rpc::JobTableData_JobState_Name(job_data.state());
+    return;
+  }
+
+  if (job_data.is_submitted_from_dashboard()) {
+    auto job_table_data = std::make_shared<rpc::JobTableData>(job_data);
+    const bool start_driver = job_data.raylet_id() == self_node_id_.Binary();
+    agent_manager_->InitializeJobEnv(job_table_data, start_driver);
+  }
```

Comment:
Maye move all the above code to the JobManager? 
NodeManager is already too large. It'd be nice to put some functionalities in sub modules.

Reply:
The `JobManager` is pretty simple, it only manages the job table data. I think it's better to call the `InitializeJobEnv` in the `NodeManager` with other job initializing logic.